### ISSUE TITLE
Expected behavior?

### ISSUE THREAD

**Issue by jaapdejong** [hashicorp/serf#366]
Hi All,

with 0.7.0 I noticed that the memory serf is using keeps growing in time.
Serf 0.6.4 has a more or less stable footprint of a little less than 6 MB.
Serf 0.7.0 starts with just over 5 MB but then grows within the hour to over 9 MB, not sure where that stops.
Is this behavior expected or am I observing a memory leak?
My test setup is only 2 serf clients.
Memory consumption is measured by looking at the value of VmRSS in /proc/<PID>/status

Cheers,
Jaap

**Comment by jaapdejong** [comment]
not sure if it matters...
Serf 0.6.4 translated with go 1.4.3
Serf 0.7.0 with 1.5.2

**Comment by slackpad** [comment]
Hi @jaapdejong we haven't observed any memory leaks but Go 1.5.2 has significantly different garbage collection behavior from 1.4.3, so a difference on the order of a few megabytes seems reasonable. Can you report back in your environment after a few days of uptime and see if things level out?

**Comment by jaapdejong** [comment]
Hi @slackpad I will. I might also test with serf 0.7.0 compiled with go 1.4.3.

**Comment by jaapdejong** [comment]
Seems that the new go1.5.2 compiler is responsible for the growth.
With go1.4.3 the memory consumption for serf0.7.0 looks same as for serf0.6.4

**Comment by jaapdejong** [comment]
Might do some experiments with serf0.7.0+go1.5.2 and setting GOGC to some interesting value...

**Comment by jaapdejong** [comment]
serf0.7.0+go1.5.2 with GOGC=30 gives about same memory consumption compared to serf0.6.4+go1.4.3 (and serf0.7.0+go1.4.3)
Not sure about the impact in terms of cpu load and performance

**Comment by shanemeyers** [comment]
I've been experiencing a memory leak with serf as well. All machines are running 0.7.0, downloaded precompiled from the main download page. Looking across a number of machines in our infrastructure, serf uses less than 20MB of memory after starting. But over time the memory usage keeps growing.

Output of `ps -C serf -o cmd,rss,vsz,start` across all machines:
```
CMD                           RSS    VSZ  STARTED
/usr/bin/serf agent -config 15988  23916 06:16:49
/usr/bin/serf agent -config 16024  22896 06:03:10
/usr/bin/serf agent -config 16164  24972 02:04:07
/usr/bin/serf agent -config 16180  22896 05:03:51
/usr/bin/serf agent -config 16196  23916 03:20:53
/usr/bin/serf agent -config 16244  22896 06:03:08
/usr/bin/serf agent -config 16308  22896 07:01:03
/usr/bin/serf agent -config 16340  24972 02:19:27
----- trimmed -----
/usr/bin/serf agent -config 975328 1020072   Mar 08
/usr/bin/serf agent -config 985064 1058904   Mar 08
/usr/bin/serf agent -config 985884 1066536   Mar 11
/usr/bin/serf agent -config 988564 1066572   Apr 26
/usr/bin/serf agent -config 993980 1076216   Mar 07
/usr/bin/serf agent -config 994560 1061688   Mar 08
/usr/bin/serf agent -config 994652 1035516   Mar 08
```

What can be done to fix the memory leak? In the meantime I'm going to restart serf periodically, but I don't like having to do that if I can avoid it.

Serf config:
```
{
    "node_name": "foo",
    "tags": {
        "env": "foo",
        "tier": "bar"
    },
    "event_handlers": [
        "/etc/serf/handler.sh"
    ],
    "start_join": [
        "host1.example.net",
        "host2.example.net"
    ],
    "snapshot_path": "/var/lib/serf/snapshot",
    "enable_syslog": true,
    "log_level": "warn"
}
```

**Comment by slackpad** [comment]
Hi @shanemeyers it's likely https://github.com/hashicorp/serf/pull/381 - sorry I should have linked these two issues. Can you try a build with that change and see if it fixes your issue?

**Comment by shanemeyers** [comment]
@slackpad is there prebuilt binaries available? Or will the next release of serf be soon? I don't want to setup the Go env if I don't need to.

**Comment by slackpad** [comment]
Hi @shanemeyers, unfortunately we haven't released a Serf build that picks up this change, and the next release is several weeks away.

**Comment by slackpad** [comment]
Should be fixed in 0.8!
